File: Factory Method/Exemplos/Ex1_Veiculos/Ex3.java
/**
 * PADRÃO: FACTORY METHOD (MÉTODO FÁBRICA)
 * 
 * O Factory Method define uma interface para criar um objeto, mas deixa 
 * as subclasses decidirem qual classe instanciar. Ele permite que uma 
 * classe adie a instanciação para as subclasses.
 */

// 1. PRODUTO ABSTRATO (Interface ou Classe Base)
abstract class Veiculo {
    protected String modelo;

    public Veiculo(String modelo) {
        this.modelo = modelo;
    }

    public void mostrarDetalhes() {
        System.out.printf("Veículo Modelo: %s\n", this.modelo);
    }
}

// 2. PRODUTOS CONCRETOS (Implementações específicas)
class Carro extends Veiculo {
    public Carro(String modelo) {
        super(modelo);
    }
}

class Moto extends Veiculo {
    public Moto(String modelo) {
        super(modelo);
    }
}

// 3. CRIADOR ABSTRATO (Define o Factory Method)
abstract class FabricaDeVeiculos {
    // Este é o Factory Method
    public abstract Veiculo criarVeiculo(String modelo);
}

// 4. CRIADORES CONCRETOS (Sobrescrevem o Factory Method)
class FabricaDeCarros extends FabricaDeVeiculos {
    @Override
    public Veiculo criarVeiculo(String modelo) {
        return new Carro(modelo);
    }
}

class FabricaDeMotos extends FabricaDeVeiculos {
    @Override
    public Veiculo criarVeiculo(String modelo) {
        return new Moto(modelo);
    }
}

// 5. CLIENTE
public class Ex3 {
    public static void main(String[] args) {
        // Criando as fábricas
        FabricaDeVeiculos fabricaCarros = new FabricaDeCarros();
        FabricaDeVeiculos fabricaMotos = new FabricaDeMotos();

        // Usando as fábricas para criar instâncias sem conhecer as classes concretas
        Veiculo v1 = fabricaCarros.criarVeiculo("Sedan");
        Veiculo v2 = fabricaMotos.criarVeiculo("Esportiva");

        v1.mostrarDetalhes();
        v2.mostrarDetalhes();
    }
}

File: Factory Method/Exemplos/Ex1_Veiculos/Ex3.py
"""
PADRÃO: FACTORY METHOD (MÉTODO FÁBRICA)

O Factory Method permite que uma classe delegue a responsabilidade de 
instanciação para suas subclasses, promovendo o desacoplamento.
"""

from abc import ABC, abstractmethod

# 1. PRODUTO ABSTRATO
class Veiculo(ABC):
    def __init__(self, modelo):
        self.modelo = modelo

    def mostrar_detalhes(self):
        print(f"Veículo Modelo: {self.modelo}")

# 2. PRODUTOS CONCRETOS
class Carro(Veiculo):
    pass

class Moto(Veiculo):
    pass

# 3. CRIADOR ABSTRATO
class FabricaDeVeiculos(ABC):
    @abstractmethod
    def criar_veiculo(self, modelo) -> Veiculo:
        """Este é o Factory Method."""
        pass

# 4. CRIADORES CONCRETOS
class FabricaDeCarros(FabricaDeVeiculos):
    def criar_veiculo(self, modelo):
        return Carro(modelo)

class FabricaDeMotos(FabricaDeVeiculos):
    def criar_veiculo(self, modelo):
        return Moto(modelo)

# 5. CLIENTE
if __name__ == "__main__":
    fabrica_carros = FabricaDeCarros()
    fabrica_motos = FabricaDeMotos()

    v1 = fabrica_carros.criar_veiculo("Hatchback")
    v2 = fabrica_motos.criar_veiculo("Touring")

    v1.mostrar_detalhes()
    v2.mostrar_detalhes()
